### DISCUSSION TITLE
Trace interpreted languages

### DISCUSSION THREAD

**Discussion by ghost** [bpftrace/bpftrace#2337] [Q&A]
Is possible to trace library function calls from JVM, ART (android), etc.

**Comment by viktormalik** [comment]
In general, if you have a binary and you know the name of the function inside it that you want to trace, then you can attach to it using uprobes. Otherwise no, we have no support for interpreted languages, yet.

**Comment by brendangregg** [comment]
Which library calls do you mean? I'll answer using Java as an example:

- A library of Java methods (typically a class): This requires heap tracing and we aren't there yet.
- A JVM internal library (A *.so file, written in C++): Sure. Use uprobes.
- A system/external library loaded via JNI (A *.so file, written in C/C++/etc.): Usually yes, use uprobes.

(I go through examples in my BPF book if you have a copy.)
